### PR TITLE
not submit metrics if measured is stopped

### PR DESCRIPTION
**Pull request by fffw**
@uaalto mind have a look?

### CODE REVIEW

File: measured.go [hunk 1]
```diff
@@ -140,16 +140,21 @@ func Listener(l net.Listener, interval time.Duration) *MeasuredListener {
...
+		stopped: 1,
 	}
 }
 
 // Start runs the measured loop
 // Reporting interval should be same for all reporters, as cached data should
 // be cleared after each round.
 func (m *Measured) Start(reportInterval time.Duration, reporters ...Reporter) {
-	go m.run(reportInterval, reporters...)
+	if atomic.CompareAndSwapInt32(&m.stopped, 1, 0) {
```
Comment (uaalto) [inline measured.go]: This brings restarting capabilities, right?

File: measured.go [hunk 2]
```diff
@@ -416,24 +425,20 @@ func (m *Measured) submitError(connID string, err error, phase string) {
...
 	t := &Traffic{
 		ID:       connID,
 		BytesIn:  BytesIn,
 		BytesOut: BytesOut,
 	}
-	log.Tracef("Submiting traffic %+v", t)
-	select {
-	case m.chStat <- t:
-	default:
-		log.Error("Failed to submit traffic, channel busy")
```
Comment (uaalto) [inline measured.go]: Do you think that we could keep the error only in Trace level? Is it worth?

File: measured.go [hunk 3]
```diff
@@ -140,16 +140,21 @@ func Listener(l net.Listener, interval time.Duration) *MeasuredListener {
 func New() *Measured {
 	return &Measured{
 		// to avoid blocking when busily reporting stats
-		chStat: make(chan Stat, 10),
-		chStop: make(chan struct{}),
+		chStat:  make(chan Stat, 10),
```
Comment (uaalto) [inline measured.go]: Why 10?
We've seen in high-load situations temporary errors submitting metrics. Could it be favored by this buffer being too shallow?

**Comment by uaalto** [comment]
On this machine, Go 1.5.4 Debian Linux, `go test` fails some times (about 30%). I shows some times:

```
--- FAIL: TestReportStats (0.41s)
        Location:       measured_test.go:91
        Error:          Not equal: 3 (expected)
                                != 2 (actual)

FAIL
```

And some other times:

```
DEBUG measured: measured.go:164 Stopping measured loop...
--- FAIL: TestReportStats (0.41s)
        Location:       measured_test.go:102
        Error:          Not equal: "127.0.0.1:53266" (expected)
                                != "127.0.0.1:45318" (actual)
        Messages:       client stats should report server as remote addr

        Location:       measured_test.go:103
        Error:          Not equal: 0x0 (expected)
                                != 0xb0 (actual)
        Messages:       should only report increased byte count

        Location:       measured_test.go:104
        Error:          Not equal: 0x0 (expected)
                                != 0x60 (actual)
        Messages:       should only report increased byte count

        Location:       measured_test.go:105
        Error:          Not equal: 0x0 (expected)
                                != 0xb0 (actual)
        Messages:       should only report increased byte count

        Location:       measured_test.go:106
        Error:          Not equal: 0x0 (expected)
                                != 0x60 (actual)
        Messages:       should only report increased byte count

        Location:       measured_test.go:107
        Error:          Not equal: 0x0 (expected)
                                != 0xb0 (actual)
        Messages:       should only report increased byte count

        Location:       measured_test.go:108
        Error:          Not equal: 0x0 (expected)
                                != 0x60 (actual)
        Messages:       should only report increased byte count

        Location:       measured_test.go:111
        Error:          Not equal: "127.0.0.1:45318" (expected)
                                != "127.0.0.1:53266" (actual)
        Messages:       should report server stats with client addr as ID

        Location:       measured_test.go:112
        Error:          Not equal: 0x60 (expected)
                                != 0x0 (actual)
        Messages:       should report server stats with bytes in

        Location:       measured_test.go:113
        Error:          Not equal: 0xb0 (expected)
                                != 0x0 (actual)
        Messages:       should report server stats with bytes out

        Location:       measured_test.go:114
        Error:          Not equal: 0x60 (expected)
                                != 0x0 (actual)
        Messages:       should report server stats with bytes in

        Location:       measured_test.go:115
        Error:          Not equal: 0xb0 (expected)
                                != 0x0 (actual)
        Messages:       should report server stats with bytes out

        Location:       measured_test.go:116
        Error:          Not equal: 0x60 (expected)
                                != 0x0 (actual)
        Messages:       should report server stats with bytes in

        Location:       measured_test.go:117
        Error:          Not equal: 0xb0 (expected)
                                != 0x0 (actual)
        Messages:       should report server stats with bytes out

DEBUG measured: measured.go:242 Measured loop stopped
```

**Comment by uaalto** [comment]
This might be related to a race condition that I just found in http-proxy-lantern, which prevents reporting to go live: https://github.com/getlantern/http-proxy-lantern/issues/64.

Do you have the bandwidth at the moment to take care of it or do you want me to dive in?

Thank you @fffw !!

File: measured.go (hunk 1 above)
Comment (fffw) [inline measured.go]: Yeah I think so

File: measured.go (hunk 2 above)
Comment (fffw) [inline measured.go]: I don't quite understand. With the checking of `stopped` flag, `log.Error` is removed.

File: measured.go (hunk 3 above)
Comment (fffw) [inline measured.go]: Oh the buffer is not required any more.

**Comment by fffw** [comment]
The test is very sensitive to timing and thread schedule, need to improve. It always passed on my Mac so I ignored it. Thanks for checking @uaalto !

File: measured.go [hunk 4]
```diff
@@ -55,6 +56,10 @@ type LatencyTracker struct {
 	Last      time.Duration
 }
 
+func (t *LatencyTracker) String() string {
```
Comment (fffw) [inline measured.go]: It's to make `"%+v"` on a slice of pointers showing actual data.

File: measured.go [hunk 5]
```diff
@@ -140,23 +149,28 @@ func Listener(l net.Listener, interval time.Duration) *MeasuredListener {
...
+		log.Debug("measured loop already started")
+	}
 }
 
 // Stop stops the measured loop
 func (m *Measured) Stop() {
 	if atomic.CompareAndSwapInt32(&m.stopped, 0, 1) {
 		log.Debug("Stopping measured loop...")
-		close(m.chStop)
+		m.chStop <- struct{}{}
```
Comment (fffw) [inline measured.go]: Not close `chStop` so we can restart without altering `Measured` object.

**Comment by fffw** [comment]
@uaalto Updated per your comments. Also added a test to make sure Measured is restartable. I couldn't trigger the first failed test you mentioned, both locally and on Travis. Increasing the intervals would fix it but I'm not sure. Could you post the output of `TRACE=true go test -v` when the test fails?

File: measured.go (hunk 2 above)
Comment (uaalto) [inline measured.go]: Yeah nevermind.

**Comment by uaalto** [comment]
Awesome @fffw !
There's a bunch of improvements here, so I might just merge it and find out any issue happening in http-proxy myself or come back to you with more specific bug reports otherwise.

**Pull request merged**